SQL: Handle NaN returned for aggs in case of nulls

matriv opened the pull request:
Handle `NaN` (not a number) as `NULL`, since aggregations like MAX, MIN, AVG
return `NaN` instead of `NULL` when the bucket contains only NULL values.

Fixes: #35139

matriv committed:
SQL: Handle NaN returned for aggs in case of nulls

Handle `NaN` (not a number) as `NULL`, since aggregations like MAX, MIN, AVG
return `NaN` instead of `NULL` when the bucket contains only NULL values.

Fixes: #35139

elasticmachine commented:
Pinging @elastic/es-search-aggs

costin reviewed (commented):
This looks like a left-over of https://github.com/elastic/elasticsearch/pull/34916
That is the extraction of the aggs doesn't cover the NaN case (though `MetricAggExtractor#containsValues looks like it handled the `NaN` 
It's the agg extractor/wrapper that should handle these cases at the source, not the processor (which is a downstream consumer).

matriv commented:
@costin The `MetricAggExtractor` comes afterwords, the `NaN` in this case must be processed and filtered by the painless script generated because of the `HAVING` clause. So in our case the bucket is already filtered when it calls the `MetricAggExtractor#extract`. Or am I getting something wrong?

My other idea was to introduce the `nanSafe()` method only in the `InternalSqlScriptUtils` so it's wrapping any Numeric argument before the actual function call.

matriv committed:
Revert "SQL: Handle NaN returned for aggs in case of nulls"

This reverts commit 904f55a73375e99f2e7a61465425b689ad132cee.

matriv committed:
Merge remote-tracking branch 'upstream/master' into mt/35139

matriv committed:
SQL: Handle NaN returned for aggs in case of nulls

Handle `NaN` (not a number) as `NULL`, since aggregations like MAX, MIN, AVG
return `NaN` instead of `NULL` when the bucket contains only NULL values.

Fixes: #35139

matriv commented:
@costin @astefan Implemented a new solution as suggested by @costin.

astefan commented on x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt:
```diff
@@ -15,6 +15,7 @@ class org.elasticsearch.xpack.sql.expression.function.scalar.whitelist.InternalS
   boolean nullSafeFilter(Boolean)
   double nullSafeSortNumeric(Number)
   String nullSafeSortString(Object)
+  Number nanSafeFilter(Number)
```
Can you put the elements in this group in alphabetical order, please? I think that was the original intention with the list of functions here.

costin replied:
This name is misleading - the above methods are used to wrap an expression used in filtering; this method should be used against all aggs, regardless of their position or whether the expression is a filter or not.
How about `nullAgg()`?

astefan reviewed (approved):
LGTM.

costin commented on x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec:
```diff
@@ -467,6 +467,19 @@ SELECT languages, COUNT(*) c FROM test_emp GROUP BY languages HAVING NOT COUNT(*
 SELECT languages, COUNT(*) c FROM test_emp GROUP BY languages HAVING COUNT(*) <> 1 ORDER BY languages DESC;
 
+aggWithHavingAndComparisonOnNaN1
+SELECT max(languages) AS maxl, gender FROM "test_emp" WHERE emp_no IN(10018, 10019, 10020, 10021) GROUP BY gender HAVING maxl <= 2;
```
Can you please try other aggs as well?
For example use min and max to trigger a stats agg, kurtosis for matrix, sum (which should always return zero), percentiles and percentile ranks.
If indeed the bucket selector gives us just the value, I think the proper fix should actually be inside ES...
@polyfractal this looks like another instanceof https://github.com/elastic/elasticsearch/issues/34903 - what do you think?

polyfractal replied:
Sorry, playing catch-up here.  What's the underlying aggregation structure being generated?  Aggs + bucket_selector?  Which value(s) being aggregated are NaN/null?

costin replied:
In the example above `HAVING maxl <=2` becomes a Painless script doing a bucket selector pointing to `MAX` agg with the content of (`max_agg <=2`). 
This issue appears though for any agg - when running against a null bucket, instead of returning nulls, the aggs selected by the bucket selector their default values (like in #34903). In case of Max is a `NaN` but this clearly differs from case to case.
In scripting however we have no understanding of the source so we can't determine that it's an agg since we only receive the value (`NaN`) or whatever.

I would argue this is an actual bug in ES.

costin replied:
@polyfractal ^^

polyfractal replied:
@costin Ahh, yes I see.  I agree it's related to #34903.  Bucket selector is just resolving to the internal agg object and asking for it's `value()`, which returns the primitive `double` internal state unlike xcontent which does the null conversion.

I think we can probably expose the `hasValue()` method (or whatever the mechanism ends up being) in the bucket_selector painless context so that it can be used from scripts.  I don't think it's something we can do automatically though, because of how the pipeline framework works right now.  It's a similar issue to #27377 (not quite the same, but similar root issue).  I don't think we can fix that easily until the transport client goes away, or else we'll have a big breaking change to how agg objects work (`double` to `Double` or `Optional`)

So the script might need to do a check if there is a value before actually using the value.  Something like:

```
if (hasValue(max1) && max1.value <= 2) {
   ...
}
```

Not ideal but probably the path of least resistance right now. Or maybe introduce a new `valueOrNull()` method in the context or something.

I'll make a note on #34903 about pipelines

costin reviewed (commented):
I don't think this is a complete fix unfortunately due to https://github.com/elastic/elasticsearch/issues/34903. Trying out several aggs should indicate whether that's the case or not.

matriv commented:
@elastic/es-ql

matriv commented:
Superseded by #71525
